### PR TITLE
[WIP] - 342 issue apicurio integration refactor

### PR DESCRIPTION
**Pull request by josemenvegsngular**
refactor

### CODE REVIEW

File: src/test/java/com/sngular/kloadgen/extractor/extractors/ProtobufExtractorTest.java [hunk 1]
```diff
@@ -6,6 +6,7 @@
```
Comment (AdrianLagarteraSngular) [inline src/test/java/com/sngular/kloadgen/extractor/extractors/ProtobufExtractorTest.java:6]: Please add test to check the behavior of the new extractor interface

File: src/test/java/com/sngular/kloadgen/extractor/extractors/JsonExtractorTest.java [hunk 2]
```diff
@@ -8,6 +8,7 @@
 import com.sngular.kloadgen.model.ConstraintTypeEnum;
```
Comment (AdrianLagarteraSngular) [inline src/test/java/com/sngular/kloadgen/extractor/extractors/JsonExtractorTest.java:8]: Please add test to check the behavior of the new extractor interface

File: src/test/java/com/sngular/kloadgen/extractor/extractors/AvroExtractorTest.java [hunk 3]
```diff
@@ -22,9 +22,8 @@ void testFlatPropertiesEmbeddedAvros() throws Exception {
```
Comment (AdrianLagarteraSngular) [inline src/test/java/com/sngular/kloadgen/extractor/extractors/AvroExtractorTest.java:22]: Please add test to check the behavior of the new extractor interface

File: src/test/java/com/sngular/kloadgen/extractor/SchemaExtractorTest.java [hunk 4]
```diff
@@ -99,10 +99,10 @@ void testFlatPropertiesListWithAVRO() throws IOException, RestClientException {
 
     jMeterHelperMockedStatic.when(() -> JMeterHelper.getParsedSchema(Mockito.anyString(), Mockito.any(Properties.class))).thenReturn(parsedSchema);
     jMeterContextServiceMockedStatic.when(() -> JMeterContextService.getContext().getProperties()).thenReturn(properties);
-    Mockito.doNothing().when(avroExtractor).processField(Mockito.any(Field.class), Mockito.anyList(), ArgumentMatchers.eq(true), ArgumentMatchers.eq(false));
+    Mockito.when(avroExtractor.processConfluentParsedSchema(Mockito.any(AvroSchema.class))).thenReturn(new ArrayList<>());
     schemaExtractor.flatPropertiesList("avroSubject");
 
-    Mockito.verify(avroExtractor, Mockito.times(2)).processField(Mockito.any(Field.class), Mockito.anyList(), ArgumentMatchers.eq(true), ArgumentMatchers.eq(false));
+    Mockito.verify(avroExtractor, Mockito.times(1)).processConfluentParsedSchema(Mockito.any(AvroSchema.class));
```
Comment (AdrianLagarteraSngular) [inline src/test/java/com/sngular/kloadgen/extractor/SchemaExtractorTest.java:105]: Please add a variation in each test where you test each Registry connection to avoid testing only the `processConfluentParsedSchema`

File: src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java [hunk 5]
```diff
@@ -20,25 +20,50 @@
...
+
+  @Override
+  public final List<FieldValueMapping> processConfluentParsedSchema(final Object schema) {
+    final List<FieldValueMapping> attributeList = new ArrayList<>();
+    final var protoFileElement = ((ProtobufSchema) schema).rawSchema();
+    // todo: en vez de null comprobar si hay que crear un HashMap con field.getNestedTypes()
+    protoFileElement.getTypes().forEach(field -> processField(field, attributeList, protoFileElement.getImports(), false, null));
+    return attributeList;
+  }
+
```
Comment (AdrianLagarteraSngular) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java:62]: Please check the configuration of your IDE, for some reason it doesn't put the `Override` annotation on the last method in the extractor implementation

File: src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java [hunk 6]
```diff
@@ -20,25 +20,50 @@
...
+  @Override
+  public final List<FieldValueMapping> processConfluentParsedSchema(final Object schema) {
+    final List<FieldValueMapping> attributeList = new ArrayList<>();
+    final var protoFileElement = ((ProtobufSchema) schema).rawSchema();
+    // todo: en vez de null comprobar si hay que crear un HashMap con field.getNestedTypes()
```
Comment (AdrianLagarteraSngular) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java:58]: Please remove the TODO comments, if its necessary write it in English

File: src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java [hunk 7]
```diff
@@ -20,25 +20,50 @@
...
+    final ProtoFileElement protoFileElement = ((ProtobufSchema) schema).rawSchema();
+    protoFileElement.getTypes().forEach(field -> processField(field, attributeList, protoFileElement.getImports(), true, nestedTypes));
     return attributeList;
   }
 
+  @Override
+  public final List<FieldValueMapping> processApicurioParsedSchema(final Object schema) {
+    final List<FieldValueMapping> attributeList = new ArrayList<>();
+    final var protoFileElement = ((io.apicurio.registry.utils.protobuf.schema.ProtobufSchema) schema).getProtoFileElement();
+    // todo: en vez de null comprobar si hay que crear un HashMap con field.getNestedTypes()
```
Comment (AdrianLagarteraSngular) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java:49]: Please remove the TODO comments, if its necessary write it in English

File: src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java [hunk 8]
```diff
@@ -204,31 +225,37 @@ private List<FieldValueMapping> extractMapInternalFields(
           FieldValueMapping.builder()
                            .fieldName(name)
                            .fieldType(value.getType() + "-map" + ((StringUtils.isNotEmpty(breadCrumb) && breadCrumb.endsWith("[:]")) ? "-map" :
-                               (StringUtils.isNotEmpty(breadCrumb) && breadCrumb.endsWith("[]")) ? "-array" : breadCrumb))
+                                                                                                                                                  (StringUtils.isNotEmpty(
```
Comment (AdrianLagarteraSngular) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java:228]: Check if the formatter its ok, the concatenation of this one should be in one or two lines instead of this

File: src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java [hunk 9]
```diff
@@ -175,7 +192,11 @@ private List<FieldValueMapping> extractArrayInternalFields(
             FieldValueMapping.builder()
                              .fieldName(name)
                              .fieldType(value.getType() + "-array" + ((StringUtils.isNotEmpty(breadCrumb) && breadCrumb.endsWith("[]")) ? "-array" :
-                                 (StringUtils.isNotEmpty(breadCrumb) && breadCrumb.endsWith("[:]")) ? "-map" : breadCrumb))
```
Comment (AdrianLagarteraSngular) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java:178]: Check if the formatter its ok, the concatenation of this one should be in one or two lines instead of this

File: src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java [hunk 10]
```diff
@@ -19,16 +18,33 @@
...
+  @Override
+  public final List<FieldValueMapping> processConfluentParsedSchema(final Object schema) {
+    return processSchema(((JsonSchema) schema).toJsonNode());
+  }
+
+  @Override
+  public final ParsedSchema getParsedSchema(final String schema) {
+    return new JsonSchema(schema);
+  }
+
```
Comment (AdrianLagarteraSngular) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java:45]: Please check the configuration of your IDE, for some reason it doesn't put the `Override` annotation on the last method in the extractor implementation

File: src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java (hunk 5 above)
Comment (josemenvegsngular) [reply src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java:62]: It seems to work fine, if I remove the "final" in the definition, it send and advise.
Even testing other interface implementations (like AvroSRLoadGenerator), there is no override tag and I see no advices.
Are you sure that rules is included?

File: src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java (hunk 6 above)
Comment (josemenvegsngular) [reply src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java:58]: I will translate is due to this comment come from the previous version ans I have no idea if thery are neede or not

File: src/test/java/com/sngular/kloadgen/extractor/extractors/AvroExtractorTest.java (hunk 3 above)
Comment (josemenvegsngular) [reply src/test/java/com/sngular/kloadgen/extractor/extractors/AvroExtractorTest.java:22]: the interface?

File: src/test/java/com/sngular/kloadgen/extractor/SchemaExtractorTest.java (hunk 4 above)
Comment (josemenvegsngular) [reply src/test/java/com/sngular/kloadgen/extractor/SchemaExtractorTest.java:105]: Do we have any input file in apicurio format?

File: src/test/java/com/sngular/kloadgen/extractor/extractors/AvroExtractorTest.java (hunk 3 above)
Comment (AdrianLagarteraSngular) [reply src/test/java/com/sngular/kloadgen/extractor/extractors/AvroExtractorTest.java:22]: Yes, when I refer to the interface on that line it is for the methods related to the new `Extractor` interface instead of call the missing test methods

Comment (josemenvegsngular) [reply src/test/java/com/sngular/kloadgen/extractor/extractors/AvroExtractorTest.java:22]: Do we have any input example? because that code hadn't got test before the refactor, and I have no idea of the input needed.

File: src/test/java/com/sngular/kloadgen/extractor/SchemaExtractorTest.java (hunk 4 above)
Comment (AdrianLagarteraSngular) [reply src/test/java/com/sngular/kloadgen/extractor/SchemaExtractorTest.java:105]: Looks like you need create your own file for this purpose

Comment (josemenvegsngular) [reply src/test/java/com/sngular/kloadgen/extractor/SchemaExtractorTest.java:105]: And I have no idea how that file looks like.
So, I open to suggestions.

Comment (josemenvegsngular) [reply src/test/java/com/sngular/kloadgen/extractor/SchemaExtractorTest.java:105]: I have had to disable the new tests due to a cast issue during the excution after a merge: I don't know if the problem is in the code or in the tests.

File: src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java [hunk 11]
```diff
@@ -19,16 +18,33 @@
...
-public class JsonExtractor {
+public class JsonExtractor implements Extractor {
 
   private final JSONSchemaParser jsonSchemaParser = new JSONSchemaParser();
 
-  public final List<FieldValueMapping> processSchema(final JsonNode jsonNode) {
-    return processSchema(jsonSchemaParser.parse(jsonNode));
+  @Override
+  public final List<FieldValueMapping> processApicurioParsedSchema(final Object schema) {
+    return processSchema(((io.apicurio.registry.serde.jsonschema.JsonSchema) schema).toJsonNode());
```
Comment (davidgayoso) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java:33]: If it is possible, please make an import of io.apicurio.registry.serde.jsonschema.JsonSchema.

File: src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java [hunk 12]
```diff
@@ -12,31 +12,26 @@
...
 import com.sngular.kloadgen.model.FieldValueMapping;
+import com.sngular.kloadgen.sampler.schemaregistry.schema.ApicurioParsedSchema;
 import com.sngular.kloadgen.util.JMeterHelper;
-import com.squareup.wire.schema.internal.parser.TypeElement;
+import com.sngular.kloadgen.util.SchemaRegistryKeyHelper;
 import io.confluent.kafka.schemaregistry.ParsedSchema;
-import io.confluent.kafka.schemaregistry.avro.AvroSchema;
 import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException;
-import io.confluent.kafka.schemaregistry.json.JsonSchema;
-//import io.confluent.kafka.schemaregistry.protobuf.ProtobufSchema;
```
Comment (davidgayoso) [inline src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java:34]: Please remove this import if you are not using it.

File: src/test/java/com/sngular/kloadgen/extractor/extractors/AvroExtractorTest.java [hunk 13]
```diff
@@ -38,17 +29,30 @@ void testFlatPropertiesEmbeddedAvros() throws Exception {
                   FieldValueMapping.builder().fieldName("timestamp").fieldType("long").fieldValueList("").required(true).isAncestorRequired(true).build()
               );
   }
+  @Test
+  @DisplayName("Should extract Embedded Record")
+  void testFlatPropertiesEmbeddedAvros() throws Exception {
+
+    final String testFile = fileHelper.getContent("/avro-files/embedded-avros-example-test.avsc");
```
Comment (davidgayoso) [inline src/test/java/com/sngular/kloadgen/extractor/extractors/AvroExtractorTest.java:36]: I think you are using this file several times in the test, so maybe is a good idea to create a constant to hold that value

**Comment by 5uso-sng** [comment]
Since #349 was already merged into main, does it make sense for this PR to merge into branch 302?

File: src/main/java/com/sngular/kloadgen/extractor/extractors/AvroExtractor.java [hunk 14]
```diff
@@ -14,15 +14,30 @@
...
+  @Override
+  public final List<FieldValueMapping> processSchema(final Object schema) {
     final var attributeList = new ArrayList<FieldValueMapping>();
-    schema.getFields().forEach(field -> processField(field, attributeList, true, true));
+    ((AvroSchema) schema).rawSchema().getFields().forEach(field -> processField(field, attributeList, true, true));
+    return attributeList;
+  }
+
+  @Override
+  public final List<FieldValueMapping> processApicurioParsedSchema(final Object schema) {
```
Comment (rabad-sng) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/AvroExtractor.java:31]: The methods processApicurioParsedSchema and processConvluenteParsedSchema have the same implementation except for the type of argument. To avoid casting and because both methods should be technology provider agnostic, I suggest to specialise the input argument called schema and unify both implementations.

File: src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java [hunk 15]
```diff
@@ -19,16 +18,33 @@
...
 
-public class JsonExtractor {
+public class JsonExtractor implements Extractor {
 
   private final JSONSchemaParser jsonSchemaParser = new JSONSchemaParser();
 
-  public final List<FieldValueMapping> processSchema(final JsonNode jsonNode) {
-    return processSchema(jsonSchemaParser.parse(jsonNode));
+  @Override
+  public final List<FieldValueMapping> processApicurioParsedSchema(final Object schema) {
```
Comment (rabad-sng) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java:32]: Same as above comments

File: src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java [hunk 16]
```diff
@@ -60,47 +55,48 @@ public SchemaExtractorImpl(final AvroExtractor avroExtractor, final JsonExtracto
     this.protoBufExtractor = protoBufExtractor;
   }
 
+  private Extractor getExtractor(final String schemaType) {
```
Comment (rabad-sng) [inline src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java:58]: I'd move this block of code to a new Factory Method

File: src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java [hunk 17]
```diff
@@ -60,47 +55,48 @@ public SchemaExtractorImpl(final AvroExtractor avroExtractor, final JsonExtracto
...
-      } else if (SchemaTypeEnum.PROTOBUF.name().equalsIgnoreCase(schemaType)) {
-        final var protoFileElement = ((io.apicurio.registry.utils.protobuf.schema.ProtobufSchema) schema).getProtoFileElement();
-        // todo: en vez de null comprobar si hay que crear un HashMap con field.getNestedTypes()
-        protoFileElement.getTypes().forEach(field -> protoBufExtractor.processField(field, attributeList, protoFileElement.getImports(), false, null));
-      } else {
-        throw new KLoadGenException(String.format("Schema type not supported %s", apicurioParsedSchema.getType()));
-      }
-
+      attributeList = getExtractor(schemaType).processApicurioParsedSchema(schema);
     } else if (Objects.nonNull(registryName) && registryName.equalsIgnoreCase(SchemaRegistryEnum.CONFLUENT.toString())) {
```
Comment (rabad-sng) [inline src/main/java/com/sngular/kloadgen/extractor/impl/SchemaExtractorImpl.java:95]: We want to avoid have to add new if-else statement if new registry vendors are included. We should use Factory Method

**Review by rabad-sng — CHANGES_REQUESTED** [review]

File: src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java [hunk 18]
```diff
@@ -20,25 +20,50 @@
...
+    final ProtoFileElement protoFileElement = ((ProtobufSchema) schema).rawSchema();
+    protoFileElement.getTypes().forEach(field -> processField(field, attributeList, protoFileElement.getImports(), true, nestedTypes));
     return attributeList;
   }
 
+  @Override
+  public final List<FieldValueMapping> processApicurioParsedSchema(final Object schema) {
+    final List<FieldValueMapping> attributeList = new ArrayList<>();
+    final var protoFileElement = ((io.apicurio.registry.utils.protobuf.schema.ProtobufSchema) schema).getProtoFileElement();
+    // TODO: instead of sending null, check if it's needed to create an HashMap with field.getNestedTypes()
```
Comment (5uso-sng) [inline src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java:49]: Is there a reason why this change can't be done along with the rest of the PR?

**Review by 5uso-sng — COMMENTED** [review]
Nothing seems out of place

File: src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java (hunk 11 above)
Comment (josemenvegsngular) [reply src/main/java/com/sngular/kloadgen/extractor/extractors/JsonExtractor.java:33]: it's not possible, I have laready imported `io.confluent.kafka.schemaregistry.json.JsonSchema`

File: src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java (hunk 18 above)
Comment (josemenvegsngular) [reply src/main/java/com/sngular/kloadgen/extractor/extractors/ProtoBufExtractor.java:49]: It's an old comment and I have no idea of what it speaks about

File: src/main/java/com/sngular/kloadgen/extractor/extractors/AvroExtractor.java (hunk 14 above)
Comment (josemenvegsngular) [reply src/main/java/com/sngular/kloadgen/extractor/extractors/AvroExtractor.java:31]: ¿?
